src/Oracle/Oracle.cs

```diff
@@ -0,0 +1,464 @@
+            if (FinishedCache.Contains(requestId)) throw new RpcException(-100, "Request has already finished");
+
+            using (SnapshotView snapshot = Blockchain.Singleton.GetSnapshot())
+            {
+                if (NativeContract.Oracle.GetRequest(snapshot, requestId) is null)
+                    throw new RpcException(-100, "Request is not found");
+                AddResponseTxSign(snapshot, requestId, oraclePub, txSign);
+            }
+            return new JObject();
+        }
+
+        [ConsoleCommand("start oracle", Category = "Oracle", Description = "Start oracle service")]
+        private void OnStart()
+        {
+            string password = Helper.ReadUserInput("password", true);
+            if (password.Length == 0)
+            {
+                Console.WriteLine("Cancelled");
+                return;
+            }
+            try
+            {
+                Wallet.Unlock(password);
+            }
+            catch (System.Security.Cryptography.CryptographicException)
+            {
+                Console.WriteLine($"Failed to open wallet");
+                return;
+            }
+
+            using (var snapshot = Blockchain.Singleton.GetSnapshot())
+            {
+                if (!CheckOracleAvaiblable(snapshot, out ECPoint[] oracles)) throw new ArgumentException("The oracle service is unavailable");
+                if (!CheckOracleAccount(oracles)) throw new ArgumentException("There is no oracle account in wallet");
+            }
+            Interlocked.Exchange(ref CancelSource, new CancellationTokenSource())?.Cancel();
+            new Task(() =>
+            {
+                while (CancelSource?.IsCancellationRequested == false)
+                {
+                    using (var snapshot = Blockchain.Singleton.GetSnapshot())
+                    {
+                        IEnumerator<(ulong RequestId, OracleRequest Request)> enumerator = NativeContract.Oracle.GetRequests(snapshot).GetEnumerator();
+                        while (enumerator.MoveNext() && !CancelSource.IsCancellationRequested)
+                            if (!FinishedCache.Contains(enumerator.Current.RequestId) && (!PendingQueue.TryGetValue(enumerator.Current.RequestId, out OracleTask task) || task.Tx is null))
+                                ProcessRequest(snapshot, enumerator.Current.RequestId, enumerator.Current.Request);
+                    }
+                    Thread.Sleep(500);
+                }
+            }).Start();
+        }
+
+        void OnPersist(StoreView snapshot, IReadOnlyList<Blockchain.ApplicationExecuted> applicationExecutedList)
+        {
+            if (!CheckOracleAvaiblable(snapshot, out ECPoint[] oracles) || !CheckOracleAccount(oracles))
+                OnStop();
+        }
+
+        private bool CheckOracleAvaiblable(StoreView snapshot, out ECPoint[] oracles)
+        {
+            oracles = NativeContract.Designate.GetDesignatedByRole(snapshot, Role.Oracle, snapshot.Height);
```

Comment:
should be snapshot.Height + 1, because snapshot.Height does not include the latest height designatedRole.  Refs: https://github.com/neo-project/neo/blob/62a3ca2eb2dc7b248be1b83ac055bd46eba690d3/src/neo/SmartContract/Native/Designate/DesignateContract.cs#L33-L38

Reply:
Done